File: Winter2022/examples/java/Lecture 6/NetworkingURL/app/src/main/java/mobiledev/unb/ca/networkingurl/HttpGetTask.java
package mobiledev.unb.ca.networkingurl;

import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import java.io.BufferedInputStream;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.concurrent.Executors;

public class HttpGetTask {
    private static final String TAG = "HttpGetTask";

    // Get your own user name at http://www.geonames.org/login
    private static final String USER_NAME = "aporter";

    private static final String HOST = "api.geonames.org";
    private static final String URL = "http://" + HOST + "/earthquakesJSON?north=44.1&south=-9.9&east=-22.4&west=55.2&username="
            + USER_NAME;

    private final RetainedFragment listenerFragment;

    public HttpGetTask(RetainedFragment listenerFragment) {
        this.listenerFragment = listenerFragment;
    }

    public void execute() {
        // Perform background call to read the information from the URL
        Executors.newSingleThreadExecutor().execute(() -> {
            Handler handler = new Handler(Looper.getMainLooper());
            String jsonString = loadJsonFromUrl();

            handler.post(() -> updateDisplay(jsonString));
        });
    }

    private void updateDisplay(String jsonString) {
        listenerFragment.onDownloadFinished(jsonString);
    }

    private String loadJsonFromUrl() {
        HttpURLConnection httpUrlConnection = null;

        try {
            // Get the connection and prepare the request
            httpUrlConnection = (HttpURLConnection) new URL(URL).openConnection();

            // Process the response
            InputStream in = new BufferedInputStream(httpUrlConnection.getInputStream());
            return readStream(in);
        } catch (MalformedURLException exception) {
            Log.e(TAG, "MalformedURLException");
        } catch (IOException exception) {
            Log.e(TAG, "IOException");
        } finally {
            if (null != httpUrlConnection) {
                // 5. Disconnect
                httpUrlConnection.disconnect();
            }
        }

        return null;
    }

    private String readStream(InputStream in) {
        StringBuilder data = new StringBuilder();
        try (BufferedReader reader = new BufferedReader(new InputStreamReader(in))){
            String line;
            while ((line = reader.readLine()) != null) {
                data.append(line);
            }
        } catch (IOException e) {
            Log.e(TAG, "IOException");
        }

        return data.toString();
    }
}

File: Winter2022/examples/kotlin/Lecture 6/NetworkingURL/app/src/main/java/mobiledev/unb/ca/networkingurl/HttpGetTask.kt
package mobiledev.unb.ca.networkingurl

import android.os.AsyncTask
import android.util.Log
import java.io.*
import java.lang.StringBuilder
import java.lang.ref.WeakReference
import java.net.HttpURLConnection
import java.net.MalformedURLException
import java.net.URL

internal class HttpGetTask(retainedFragment: RetainedFragment?) :
    AsyncTask<Void?, Void?, String?>() {
    private val mListener: WeakReference<RetainedFragment?> = WeakReference(retainedFragment)

    override fun doInBackground(vararg params: Void?): String? {
        var httpUrlConnection: HttpURLConnection? = null
        try {
            // 1. Get connection. 2. Prepare request (URI)
            httpUrlConnection = URL(URL).openConnection() as HttpURLConnection

            // 3. This app does not use a request body
            // 4. Read the response
            val `in`: InputStream = BufferedInputStream(httpUrlConnection.inputStream)
            return readStream(`in`)
        } catch (exception: MalformedURLException) {
            Log.e(TAG, "MalformedURLException")
        } catch (exception: IOException) {
            Log.e(TAG, "IOException")
        } finally {
            httpUrlConnection?.disconnect()
        }
        return null
    }

    override fun onPostExecute(result: String?) {
        if (null != mListener.get()) {
            mListener.get()!!.onDownloadFinished(result)
        }
    }

    private fun readStream(`in`: InputStream): String {
        val data = StringBuilder()
        try {
            BufferedReader(InputStreamReader(`in`)).use { reader ->
                var line: String?
                while (reader.readLine().also { line = it } != null) {
                    data.append(line)
                }
            }
        } catch (e: IOException) {
            Log.e(TAG, "IOException")
        }
        return data.toString()
    }

    companion object {
        private const val TAG = "HttpGetTask"

        // Get your own user name at http://www.geonames.org/login
        private const val USER_NAME = "aporter"
        private const val HOST = "api.geonames.org"
        private const val URL =
            ("http://" + HOST + "/earthquakesJSON?north=44.1&south=-9.9&east=-22.4&west=55.2&username="
                    + USER_NAME)
    }

}